src/isl.rs

```diff
@@ -0,0 +1,211 @@
+            None => None // If the type is UNNAMED_TYPE_DEFINITION/ AnonymousType then set name as None
+        };
+
+        // parses all the constraints inside a Type
+        for (field_name, value) in ion_struct.iter() {
+            let constraint_name = match field_name.text() {
+                Some("name") => continue, // if the field_name is "name" then it's the type name not a constraint
+                Some(name) => name,
+                None => {
+                    return Err(invalid_schema_error_raw(
+                        "A type name symbol token does not have any text",
+                    ))
+                }
+            };
+            // TODO: add more constraints to match below
+            let constraint = match constraint_name {
+                "all_of" => {
+                    if value.ion_type() != IonType::List {
+                        return Err(invalid_schema_error_raw(format!(
+                            "all_of constraint was a {:?} instead of a list",
+                            value.ion_type()
+                        )));
+                    }
+                    let types: Vec<IslTypeRef> = value
+                        .as_sequence()
+                        .unwrap()
+                        .iter()
+                        .map(|e| IslTypeRef::parse_from_ion_element(e))
+                        .collect::<IonSchemaResult<Vec<IslTypeRef>>>()?;
+                    IslConstraint::AllOf(types)
+                }
+                "type" => {
+                    if value.ion_type() != IonType::Symbol && value.ion_type() != IonType::Struct {
+                        return Err(invalid_schema_error_raw(format!(
+                            "type constraint was a {:?} instead of a symbol/struct",
+                            value.ion_type()
+                        )));
+                    }
+                    let type_reference: IslTypeRef = IslTypeRef::parse_from_ion_element(value)?;
+                    IslConstraint::Type(type_reference)
+                }
+                _ => {
+                    return Err(invalid_schema_error_raw(
+                        "Type: ".to_owned()
+                            + &type_name.unwrap()
+                            + " can not be built as constraint: "
+                            + constraint_name
+                            + " does not exist",
+                    ))
+                }
+            };
+            constraints.push(constraint);
+        }
+        Ok(IslType::new(type_name.to_owned(), constraints))
+    }
+}
+
+/// Provides an internal representation of schema type reference.
+/// Type reference grammar is defined in [Ion Schema Spec]
+/// [Ion Schema spec]: https://amzn.github.io/ion-schema/docs/spec.html#grammar
+#[derive(Debug, Clone)]
+pub enum IslTypeRef {
+    /// represents core ion type reference
+    IslCoreType(IonType),
+    /// represents a type reference which represents a type imported from another schema
```

Comment:
Consider changing this to `NamedType` since these won't always be an alias, per se.

```suggestion
    /// represents a reference to a named type (including aliases)
```